enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,664 @@
+
+See the API Extensions section below for sample install-configs.
+
+For a two-node cluster to be successful, we need to ensure the following:
+1. The BMC secrets for RHEL-HA are created on disk during bootstrapping by the OpenShift installer via a MachineConfig.
+2. When pacemaker is initialized by the in-cluster entity responsible for starting pacemaker, pacemaker will try to set up fencing with this secret. If this is not successful, it throws an error and the installation fails.
+3. Pacemaker periodically checks that the fencing agent is healthy (i.e. can connect to the BMC) and throws a warning if it cannot access the BMC. There is an open question on what the user experience should be to raise this error to the user.
+4. The cluster will continue to run normally in the state where the BMC cannot be accessed, but ignoring this warning will mean that pacemaker can only provide a best-effort recovery - so operations that require fencing will need manual recovery.
+
+#### Day 2 Procedures
+
+As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes can not be applied when the cluster is in a degraded state.
+Such operations will only proceed when both peers are online and healthy.
+
+The experience of managing a two-node control-plane should be largely indistinguishable from that of a 3-node one.
+The primary exception is (re)booting one of the peers while the other is offline and expected to remain so.
+
+As in a 3-node control-plane cluster, starting only one node is not expected to result in a functioning cluster.
+Should the admin wish for the control-plane to start, the admin will need to execute a supplied confirmation command on the active cluster node.
+This command will grant quorum to the RHEL-HA components, authorizing it to fence its peer and start etcd as a cluster-of-one in read/write mode.
+Confirmation can be given at any point and optionally make use of SSH to facilitate initiation by an external script.
+
+### API Extensions
+
+Three known capabilities require API extensions.
+
+1. Identifying two-node control-plane clusters as a unique topology
+2. Telling CEO when it is safe for it to disable certain membership-related functionalities
+3. Collecting fencing credentials for pacemaker initialization in the install-config
+
+#### Unique Topology
+
+A mechanism is needed for components of the cluster to understand that this is a two-node control-plane topology that may require different handling.
+We will define a new value for the `TopologyMode` enum: `DualReplica`.
+The enum is used for the `controlPlaneTopology` and `infrastructureTopology` fields, and the currently supported values are `HighlyAvailable`, `SingleReplica`, and `External`.
```

Comment:
I have set aside a task to go through the arbiter EP and port over any relevant/related content. It is strangely absent from this document. :sweat_smile:

Reply:
I believe the proposal was _not_ to add a new topology for clusters with arbiters?
Note that if you create a new topology then all OLM operators have to be updated to understand what to do with it.
For 2NO I think the appropriate `infrastructureTopology` is still `Highly Available`, because a regular HA cluster has a minimum of only 2 workers anyway. At least, that's the argument I made on the arbiter EP :slightly_smiling_face:
Arguably there is a need for a separate `controlPlaneTopology` here, since up to now HA has always meant 3 control plane nodes, and some operators may be relying on that. Hopefully fewer operators are looking at `controlPlaneTopology` than would be looking at `infrastructureTopology`.